R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R

```diff
@@ -75,46 +73,55 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop <- function(command, ...) {
-      # Take the FAO Harvested area data in table L101.ag_HA_bm2_R_C_Y and get the average value for each
+      # GPK note August 2018 - this scaler used to be computed at the level of GCAM region and commodity. The method
+      # is revised to instead apply by country and specific crop. This addresses (further) data discrepancies on
+      # fodder crop production and harvested area that would otherwise lever off of one another to make unrealistic
+      # fertilizer IO coefficients in selected region/crop/GLUs.
+
+      # First, calculate the FAO_LDS scaler for each nation-crop combination
+      # Take the FAO Harvested area data in table L100.FAO_ag_HA_ha and get the average value for each
       # region-commodity combination over the Monfreda/LDS years = 1998-2002 default.
-      # Then Take the LDS harvested area data, L102.ag_HA_bm2_R_C_GLU, sum to the region-commodity level,
+      # Then Take the LDS harvested area data, L100.LDS_ag_HA_ha, sum to the nation-crop level,
       # join it to the averaged FAO data and compute an FAO_LDS scaler as FAO/LDS:
       #
-      # Take FAO HA data and calulate the average value over aglu.FAO_LDS_YEARS for each region-commodity combo:
-      L101.ag_HA_bm2_R_C_Y %>%
+      # Take FAO HA data and calulate the average value over aglu.FAO_LDS_YEARS for each nation-crop combo:
+      # Note - using FAO's "item" here rather than LDS's "GTAP_crop" because the latter has two crops that correspond
+      # to the FAO's "Grasses Nes for forage;Sil"
+      L100.FAO_ag_HA_ha %>%
         filter(year %in% aglu.FAO_LDS_YEARS) %>%
-        group_by(GCAM_region_ID, GCAM_commodity) %>%
-        summarise(FAO = mean(value)) ->
+        group_by(iso, item) %>%
+        summarise(FAO = mean(value)) %>%
+        ungroup() ->
         L141.FAO
       # Take LDS HA data and aggregate to the GCAM region-commodity level:
-      L102.ag_HA_bm2_R_C_GLU %>%
-        group_by(GCAM_region_ID, GCAM_commodity) %>%
-        summarise(LDS = sum(value)) %>% ungroup() %>%
-        # Join in the FAO data from the previous pipeline
-        left_join_error_no_match(L141.FAO, by = c("GCAM_region_ID", "GCAM_commodity")) %>%
+      L100.LDS_ag_HA_ha %>%
+        left_join(FAO_ag_items_PRODSTAT[c("item", "GTAP_crop")], by = "GTAP_crop") %>%
+        drop_na() %>%                 # Some of the GTAP crops don't have a corresponding crop in the FAO databases
+        group_by(iso, item) %>%
+        summarise(LDS = sum(value)) %>%
         ungroup() %>%
+        # Join in the FAO data from the previous pipeline
+        left_join(L141.FAO, by = c("iso", "item")) %>%
+        replace_na(list(FAO = 0)) %>%
         # Calculate the FAO_LDS scaler value = FAO/LDS
-        mutate(scaler = FAO / LDS) ->
+        mutate(scaler = FAO / LDS) %>%
+        # join back in the GTAP crop, which will be used for the remainder of the processing
+        left_join(FAO_ag_items_PRODSTAT[c("item", "GTAP_crop")], by = "item") %>%
```

Comment:
I tried swapping in `left_join_error_no_match()` but it did get:
```
Error in left_join_error_no_match(., FAO_ag_items_PRODSTAT[c("item", "GTAP_crop")],  : 
  left_join_no_match: number of rows in data changed 
```

@pkyle, just wanted to make sure this is expected.  I did notice a few lines up:
```
# Some of the GTAP crops don't have a corresponding crop in the FAO databases
```
But again, figure better to double check.

Reply:
ljenm fails because there are two "GTAP crops" assigned to the FAO commodity "Grasses for forage;sil". So, the FAO commodity is listed twice in the mapping file. Fodder crops are a mess, because (1) FAO was guessing most of the data anyway, and (2) FAO slow-walked the phase-out of their reporting. As of 2017 they're totally scrubbed from FAOSTAT, but there were non-trivial differences both in reported quantities and in commodities between when Monfreda et al queried them in ~2007 and when I did in ~2014. Incidentally, the double-assignment of this FAO commodity maybe was wrong, and instead the curious "pumpkins for fodder" commodity should have been used. It's hard to tell.